File: src/main/java/com/yuyan/imemodule/utils/expression/shuntingyard/ShuntingYard.java
package com.yuyan.imemodule.utils.expression.shuntingyard;

import com.yuyan.imemodule.utils.expression.function.Function;
import com.yuyan.imemodule.utils.expression.operator.Operator;
import com.yuyan.imemodule.utils.expression.tokenizer.OperatorToken;
import com.yuyan.imemodule.utils.expression.tokenizer.Token;
import com.yuyan.imemodule.utils.expression.tokenizer.Tokenizer;
import java.util.*;

/**
 * Shunting yard implementation to convert infix to reverse polish notation
 */
public class ShuntingYard {

    /**
     * Convert a Set of tokens from infix to reverse polish notation
     *
     * @param expression             the expression to convert
     * @param userFunctions          the custom functions used
     * @param userOperators          the custom operators used
     * @param variableNames          the variable names used in the expression
     * @param implicitMultiplication set to false to turn off implicit multiplication
     * @return a {@link com.yuyan.imemodule.utils.expression.tokenizer.Token} array containing the result
     */
    public static Token[] convertToRPN(final String expression, final Map<String, Function> userFunctions,
                                       final Map<String, Operator> userOperators, final Set<String> variableNames, final boolean implicitMultiplication) {
        final Stack<Token> stack = new Stack<>();
        final List<Token> output = new ArrayList<>();

        final Tokenizer tokenizer = new Tokenizer(expression, userFunctions, userOperators, variableNames, implicitMultiplication);
        while (tokenizer.hasNext()) {
            Token token = tokenizer.nextToken();
            switch (token.getType()) {
                case Token.TOKEN_NUMBER:
                case Token.TOKEN_VARIABLE:
                    output.add(token);
                    break;
                case Token.TOKEN_FUNCTION:
                    stack.add(token);
                    break;
                case Token.TOKEN_SEPARATOR:
                    while (!stack.empty() && stack.peek().getType() != Token.TOKEN_PARENTHESES_OPEN) {
                        output.add(stack.pop());
                    }
                    if (stack.empty() || stack.peek().getType() != Token.TOKEN_PARENTHESES_OPEN) {
                        throw new IllegalArgumentException("Misplaced function separator ',' or mismatched parentheses");
                    }
                    break;
                case Token.TOKEN_OPERATOR:
                    while (!stack.empty() && stack.peek().getType() == Token.TOKEN_OPERATOR) {
                        OperatorToken o1 = (OperatorToken) token;
                        OperatorToken o2 = (OperatorToken) stack.peek();
                        if (o1.getOperator().getNumOperands() == 1 && o2.getOperator().getNumOperands() == 2) {
                            break;
                        } else if ((o1.getOperator().isLeftAssociative() && o1.getOperator().getPrecedence() <= o2.getOperator().getPrecedence())
                                || (o1.getOperator().getPrecedence() < o2.getOperator().getPrecedence())) {
                            output.add(stack.pop());
                        } else {
                            break;
                        }
                    }
                    stack.push(token);
                    break;
                case Token.TOKEN_PARENTHESES_OPEN:
                    stack.push(token);
                    break;
                case Token.TOKEN_PARENTHESES_CLOSE:
                    while (stack.peek().getType() != Token.TOKEN_PARENTHESES_OPEN) {
                        output.add(stack.pop());
                    }
                    stack.pop();
                    if (!stack.isEmpty() && stack.peek().getType() == Token.TOKEN_FUNCTION) {
                        output.add(stack.pop());
                    }
                    break;
                default:
                    throw new IllegalArgumentException("Unknown Token type encountered. This should not happen");
            }
        }
        while (!stack.empty()) {
            Token t = stack.pop();
            if (t.getType() == Token.TOKEN_PARENTHESES_CLOSE || t.getType() == Token.TOKEN_PARENTHESES_OPEN) {
                throw new IllegalArgumentException("Mismatched parentheses detected. Please check the expression");
            } else {
                output.add(t);
            }
        }
        return output.toArray(new Token[0]);
    }
}

File: src/main/java/com/yuyan/imemodule/libs/expression/shuntingyard/ShuntingYard.kt
package com.yuyan.imemodule.libs.expression.shuntingyard

import com.yuyan.imemodule.libs.expression.function.Function
import com.yuyan.imemodule.libs.expression.operator.Operator
import com.yuyan.imemodule.libs.expression.tokenizer.OperatorToken
import com.yuyan.imemodule.libs.expression.tokenizer.Token
import com.yuyan.imemodule.libs.expression.tokenizer.Tokenizer
import java.util.Stack

object ShuntingYard {

    fun convertToRPN(expression: String, userFunctions: Map<String, Function>, userOperators: Map<String, Operator>, variableNames: Set<String>, implicitMultiplication: Boolean): Array<Token> {
        val stack = Stack<Token>()
        val output: MutableList<Token> = ArrayList()
        val tokenizer = Tokenizer(
            expression,
            userFunctions,
            userOperators,
            variableNames,
            implicitMultiplication
        )
        while (tokenizer.hasNext()) {
            val token = tokenizer.nextToken()
            when (token.type) {
                Token.TOKEN_NUMBER, Token.TOKEN_VARIABLE -> output.add(token)
                Token.TOKEN_FUNCTION -> stack.add(token)
                Token.TOKEN_SEPARATOR -> {
                    while (!stack.empty() && stack.peek().type != Token.TOKEN_PARENTHESES_OPEN) {
                        output.add(stack.pop())
                    }
                    require(!(stack.empty() || stack.peek().type != Token.TOKEN_PARENTHESES_OPEN)) { "Misplaced function separator ',' or mismatched parentheses" }
                }

                Token.TOKEN_OPERATOR -> {
                    while (!stack.empty() && stack.peek().type == Token.TOKEN_OPERATOR) {
                        val o1 = token as OperatorToken
                        val o2 = stack.peek() as OperatorToken
                        if (o1.operator.numOperands == 1 && o2.operator.numOperands == 2) {
                            break
                        } else if (o1.operator.isLeftAssociative && o1.operator.precedence <= o2.operator.precedence || o1.operator.precedence < o2.operator.precedence) {
                            output.add(stack.pop())
                        } else {
                            break
                        }
                    }
                    stack.push(token)
                }

                Token.TOKEN_PARENTHESES_OPEN -> stack.push(token)
                Token.TOKEN_PARENTHESES_CLOSE -> {
                    while (stack.peek().type != Token.TOKEN_PARENTHESES_OPEN) {
                        output.add(stack.pop())
                    }
                    stack.pop()
                    if (!stack.isEmpty() && stack.peek().type == Token.TOKEN_FUNCTION) {
                        output.add(stack.pop())
                    }
                }
                else -> throw IllegalArgumentException("Unknown Token type encountered. This should not happen")
            }
        }
        while (!stack.empty()) {
            val t = stack.pop()
            require(!(t.type == Token.TOKEN_PARENTHESES_CLOSE || t.type == Token.TOKEN_PARENTHESES_OPEN)) { "Mismatched parentheses detected. Please check the expression" }
            output.add(t)
        }
        return output.toTypedArray<Token>()
    }
}
